### ISSUE TITLE
DevTools: Collect more info when profiling

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#16481]
Some feedback I've heard from a DevTools user (roughly transcribed by me):

> I'm trying to pinpoint those renders...with hooks, it's sometimes more unclear to me why something is rendering...I generally don't use devtools much anywhere. I use console.log. But Redux devtools worked really well for me because I could see when things were changing and what exactly changed.

Maybe we could add an opt-in mode (in Settings > Profiler) to collect more data when profiling about _why_ a component rendered. For example, if `props` or `state` changed, we could show which keys changed (just their name, not their values). Maybe we could do something similar for context and for hooks?

Then we could add this information to the right side panel for the selected fiber in the Profiler UI.

---
Originally reported via https://github.com/bvaughn/react-devtools-experimental/issues/98

**Comment by EricTheRed-VT** [comment]
The "Why did this render" section is pretty nice, but is there any way to view the values before and after?
A coworker pointed out that in the old dev tools, you could double click on an element in the Flamegraph to view its props/state, but that feature seems to be missing in the new Profiler.

**Comment by EricTheRed-VT** [comment]
This blog post demonstrates the feature in the old profiler:
https://reactjs.org/blog/2018/09/10/introducing-the-react-profiler.html

![image](https://user-images.githubusercontent.com/18130250/64190990-8458fb00-ce45-11e9-8350-081b3ac9de4f.png)

**Comment by bvaughn** [comment]
> The "Why did this render" section is pretty nice, but is there any way to view the values before and after?

No. This was only possible to do before<sup>1</sup> because DevTools eagerly serialized props/state/context and wrapped them with `immutable` objects. As you can imagine, this was pretty costly. I wrote [an overview document](https://github.com/facebook/react/blob/master/packages/react-devtools/OVERVIEW.md) about the new DevTools and the techniques it uses to dramatically improve performance. The tl;dr for this issue is that we no longer eagerly serialize values _nor_ do we wrap them in immutable structures- so it is not possible to go back and look at previous values.

My current thinking on this feature is that I don't think it will be implemented because of the overhead it would add to profiling. I opened it because it was feedback I received, and it's good to discuss. I'm pretty cautious about doing things that might significantly regress performance though.

<sup>1</sup> This wasn't possible to do before for deeply nested values either, only shallowly, because of how our "hydration" worked.

**Comment by OliverJAsh** [comment]
This would be really useful. Right now I'm investigating why there are 30 separate commits during page load when there really only needs to be a handful. The issue is that my selector's are not memoized correctly, but without seeing the props at each commit, it's impossible for me to understand which selectors are "broken" (i.e. not memoized correctly).

**Comment by mustyoshi** [comment]
I would like more than just the root components to be named in the output json. It's hard to do programmatic analysis if I can't get the name of different components.
